### ISSUE TITLE
ara.api.models.Label.MultipleObjectsReturned: get() returned more than one Label -- it returned 2!

### ISSUE THREAD

**Issue by dmsimard** [ansible-community/ara#113]
## What component is this about ?

API server

## What is happening ?

Noticed a new trace, occured with the mysql backend:
```
PLAYBOOK: smoke.yaml ***********************************************************
2020-01-30 20:51:51,886 DEBUG ara.plugins.callback.default: v2_playbook_on_start
2020-01-30 20:51:51,886 DEBUG ara.plugins.callback.default: Ignoring argument: extra_vars
2020-01-30 20:51:51,888 DEBUG urllib3.connectionpool: Starting new HTTP connection (1): localhost:53827
2020-01-30 20:51:51,948 DEBUG urllib3.connectionpool: http://localhost:53827 "POST /api/v1/playbooks HTTP/1.1" 201 1083
2020-01-30 20:51:51,949 DEBUG ara.clients.http: HTTP 201: post on /api/v1/playbooks
2020-01-30 20:51:51,949 DEBUG ara.plugins.callback.default: File not in cache, getting or creating: /home/zuul/.ara-tests/git/ara/tests/integration/smoke.yaml
2020-01-30 20:51:52,016 DEBUG urllib3.connectionpool: http://localhost:53827 "POST /api/v1/files HTTP/1.1" 201 2285
2020-01-30 20:51:52,018 DEBUG ara.clients.http: HTTP 201: post on /api/v1/files
3 plays in /home/zuul/.ara-tests/git/ara/tests/integration/smoke.yaml

PLAY [ARA Tasks test play] *****************************************************
2020-01-30 20:51:52,027 DEBUG ara.plugins.callback.default: v2_playbook_on_play_start
2020-01-30 20:51:52,096 DEBUG urllib3.connectionpool: http://localhost:53827 "PATCH /api/v1/playbooks/3 HTTP/1.1" 200 1092
2020-01-30 20:51:52,098 DEBUG ara.clients.http: HTTP 200: patch on /api/v1/playbooks/3
Internal Server Error: /api/v1/playbooks/3
Traceback (most recent call last):
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/viewsets.py", line 114, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 505, in dispatch
    response = self.handle_exception(exc)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 476, in raise_uncaught_exception
    raise exc
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 502, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/mixins.py", line 82, in partial_update
    return self.update(request, *args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/mixins.py", line 67, in update
    serializer.is_valid(raise_exception=True)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/serializers.py", line 234, in is_valid
    self._validated_data = self.run_validation(self.initial_data)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/serializers.py", line 433, in run_validation
    value = self.to_internal_value(data)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/serializers.py", line 490, in to_internal_value
    validated_value = field.run_validation(primitive_value)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/fields.py", line 565, in run_validation
    value = self.to_internal_value(data)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/relations.py", line 521, in to_internal_value
    for item in data
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/relations.py", line 521, in <listcomp>
    for item in data
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/ara/api/fields.py", line 86, in to_internal_value
    return self.get_queryset().get_or_create(**{self.slug_field: data})[0]
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/db/models/query.py", line 538, in get_or_create
    return self.get(**kwargs), False
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/db/models/query.py", line 412, in get
    (self.model._meta.object_name, num)
ara.api.models.Label.MultipleObjectsReturned: get() returned more than one Label -- it returned 2!
2020-01-30 20:51:52,183 ERROR django.request: Internal Server Error: /api/v1/playbooks/3
Traceback (most recent call last):
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/viewsets.py", line 114, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 505, in dispatch
    response = self.handle_exception(exc)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 476, in raise_uncaught_exception
    raise exc
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/views.py", line 502, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/mixins.py", line 82, in partial_update
    return self.update(request, *args, **kwargs)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/mixins.py", line 67, in update
    serializer.is_valid(raise_exception=True)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/serializers.py", line 234, in is_valid
    self._validated_data = self.run_validation(self.initial_data)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/serializers.py", line 433, in run_validation
    value = self.to_internal_value(data)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/serializers.py", line 490, in to_internal_value
    validated_value = field.run_validation(primitive_value)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/fields.py", line 565, in run_validation
    value = self.to_internal_value(data)
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/relations.py", line 521, in to_internal_value
    for item in data
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/rest_framework/relations.py", line 521, in <listcomp>
    for item in data
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/ara/api/fields.py", line 86, in to_internal_value
    return self.get_queryset().get_or_create(**{self.slug_field: data})[0]
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/db/models/query.py", line 538, in get_or_create
    return self.get(**kwargs), False
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/django/db/models/query.py", line 412, in get
    (self.model._meta.object_name, num)
ara.api.models.Label.MultipleObjectsReturned: get() returned more than one Label -- it returned 2!
2020-01-30 20:51:52,226 DEBUG urllib3.connectionpool: http://localhost:53827 "PATCH /api/v1/playbooks/3 HTTP/1.1" 500 17976
2020-01-30 20:51:52,228 ERROR ara.clients.http: Failed to patch on /api/v1/playbooks/3: {'labels': ['nodepool.provider:ovh-bhs1', 'zuul.change:693855', 'zuul.executor:ze07.openstack.org', 'zuul.job:ara-api-mysql', 'zuul.pipeline:check', 'zuul.project:opendev.org/recordsansible/ara', 'integration tests', 'smoke']}
2020-01-30 20:51:52,228 DEBUG ara.clients.http: HTTP 500: patch on /api/v1/playbooks/3
2020-01-30 20:51:52,228 ERROR ara.clients.http: Failed to patch on /api/v1/playbooks/3: {'labels': ['nodepool.provider:ovh-bhs1', 'zuul.change:693855', 'zuul.executor:ze07.openstack.org', 'zuul.job:ara-api-mysql', 'zuul.pipeline:check', 'zuul.project:opendev.org/recordsansible/ara', 'integration tests', 'smoke']}
[WARNING]: Failure using method (v2_playbook_on_play_start) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at
0x7f447a905f98>): Expecting value: line 1 column 1 (char 0)
Callback Exception: 
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/ansible/executor/task_queue_manager.py", line 324, in send_callback
    method(*new_args, **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib/python3.6/site-packages/ara/plugins/callback/ara_default.py", line 232, in v2_playbook_on_play_start
    self._set_playbook_labels(labels=labels)
   File "/home/zuul/.ara-tests/virtualenv/lib/python3.6/site-packages/ara/plugins/callback/ara_default.py", line 333, in _set_playbook_labels
    self.playbook = self.client.patch("/api/v1/playbooks/%s" % self.playbook["id"], labels=labels)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/ara/clients/http.py", line 112, in patch
    return self._request("patch", endpoint, **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/ara/clients/http.py", line 106, in _request
    return response.json()
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.6/site-packages/requests/models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
   File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
   File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
   File "/usr/lib64/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
```

## What should be happening ?

The exception should be handled.

**Comment by dmsimard** [comment]
The exception raises from this field: https://github.com/ansible-community/ara/blob/19979f9bfbfa7146cf26105338761c0fce02e72c/ara/api/fields.py#L78-L88

Which is only used once in the playbook serializer:
https://github.com/ansible-community/ara/blob/19979f9bfbfa7146cf26105338761c0fce02e72c/ara/api/serializers.py#L378-L386

The reason why ``CreatableSlugRelatedField`` exists is so the creation of the label isn't explicitely required before applying it to a playbook. It is intended to do a "get_or_create" so the client doesn't need to create the label first if it doesn't exist.

There shouldn't ever be two labels with the same name but it isn't currently enforced by the database model. Unless mistaken, I've so far only seen this issue randomly reproduce in MySQL integration test jobs from time to time.
The job runs multiple playbooks simultaneously so there might be a concurrency/race condition where a first playbook checks if the label exists, it doesn't so it gets created by while this is happening, another playbook checks and the label doesn't exist yet and so creates it.

A good first step might be to enforce the uniqueness of the name in the labels table and see if this still reproduces.

**Comment by dmsimard** [comment]
Patch for setting label names unique: https://review.opendev.org/#/c/705504/

**Comment by dmsimard** [comment]
Fixed and will be released in 1.4.
